src/sklearn_raster/ufunc.py

```diff
@@ -0,0 +1,243 @@
+            is fully masked and `skip_nodata=True`, dummy values (`nan_fill` or 0) will
+            be inserted to ensure this number of samples. This is necessary for
+            functions that require non-empty array inputs, like some scikit-learn
+            `predict` methods. No effect if the array contains enough valid pixels or if
+            `skip_nodata=False`.
+        allow_cast : bool, default False
+            If True and the `func` output dtype is incompatible with the chosen
+            `nodata_output` value, the output will be cast to the correct dtype.
+            Otherwise, an error will be raised.
+        **kwargs : dict
+            Additional keyword arguments passed to `func`.
+        """
+        # Fill NaNs in the input array if they're not being skipped
+        flat_array = self.flat_array if skip_nodata else self._fill_flat_nans(nan_fill)
+
+        # Only skip NoData if there's something to skip
+        if skip_nodata and self._num_masked > 0:
+            flat_result = self._skip_nodata_apply(
+                func,
+                flat_array=flat_array,
+                ensure_min_samples=ensure_min_samples,
+                nodata_output=nodata_output,
+                allow_cast=allow_cast,
+                nan_fill=nan_fill,
+                **kwargs,
+            )
+            # NoData is now pre-masked
+            mask_nodata = False
+        else:
+            flat_result = func(flat_array, **kwargs)
+            mask_nodata = self._num_masked > 0
+
+        @map_function_over_tuples
+        def _unflatten_and_mask(result: NDArray) -> NDArray:
+            """Unflattening result to (y, x, band) and mask NoData."""
+            output_shape = [*self.array.shape[:2], -1]
+            if mask_nodata:
+                result = self._mask_nodata(
+                    result, nodata_output=nodata_output, allow_cast=allow_cast
+                )
+
+            return result.reshape(output_shape)
+
+        return _unflatten_and_mask(flat_result)
+
+    def _skip_nodata_apply(
+        self,
+        func: ArrayUfunc,
+        *,
+        flat_array: NDArray,
+        nodata_output: float | int,
+        ensure_min_samples: int,
+        allow_cast: bool,
+        nan_fill: float | int | None,
+        **kwargs,
+    ) -> NDArray | tuple[NDArray, ...]:
+        """
+        Apply a function to all non-NoData values in a flat array.
+        """
+        if inserted_dummy_values := self._num_unmasked < ensure_min_samples:
+            if ensure_min_samples > flat_array.shape[0]:
+                raise ValueError(
+                    f"Cannot ensure {ensure_min_samples} samples with only "
+                    f"{flat_array.shape[0]} total pixels in the image chunk."
+                )
+
+            cast(NDArray, self.nodata_mask)[:ensure_min_samples] = False
+            flat_array[:ensure_min_samples] = nan_fill if nan_fill is not None else 0
```

Comment:
This could replace valid pixels if `ensure_min_samples>1`. I probably need to get the indexes of which samples in this range are NoData and use that to unmask, fill, and re-mask.

Reply:
Should be fixed by c9fa2e2